// tguess/lib/components/tguess.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const crypt_1 = require("../crypt");
const vm_1 = __importDefault(require("vm"));
function generateTGuess(dapibScriptPayload, data) {
    const { sessionToken, guess } = data;
    const [tokenKey, tokenValue] = sessionToken.split("."), preparedGuess = guess.map((entry) => ({
        ...entry,
        [tokenKey]: tokenValue,
    }));
    return new Promise((resolve, reject) => {
        const dapibReceive = (data) => {
            if (!data.tanswer) {
                return reject(`No tanswer present, there was an error during generation: ${data}`);
            }
            const tguess = data.tanswer.map((item) => Object.entries(item).reduce((newTGuess, [key, value]) => {
                newTGuess[key] = value.slice(0, -1);
                return newTGuess;
            }, {}));
            return resolve((0, crypt_1.encrypt)(JSON.stringify(tguess), sessionToken));
        };
        const virtualIframeContext = vm_1.default.createContext({
            window: {
                document: {
                    hidden: false,
                },
                parent: {
                    ae: {
                        dapibReceive,
                        answer: preparedGuess,
                    },
                },
            },
        });
        vm_1.default.runInContext(dapibScriptPayload, virtualIframeContext);
    });
}
exports.default = generateTGuess;
